soroban-sdk/src/address.rs

```diff
@@ -1,13 +1,208 @@
+
+    fn try_from_val(env: &Env, val: &Object) -> Result<Self, Self::Error> {
+        if val.is_obj_type(ScObjectType::Address) {
+            Ok(unsafe { Address::unchecked_new(env.clone(), *val) })
+        } else {
+            Err(ConversionError {})
+        }
+    }
+}
+
+impl TryFromVal<Env, RawVal> for Address {
+    type Error = <Address as TryFromVal<Env, Object>>::Error;
+
+    fn try_from_val(env: &Env, val: &RawVal) -> Result<Self, Self::Error> {
+        <_ as TryFromVal<_, Object>>::try_from_val(env, &val.try_into()?)
+    }
+}
+
+impl TryFromVal<Env, Address> for RawVal {
+    type Error = ConversionError;
+
+    fn try_from_val(_env: &Env, v: &Address) -> Result<Self, Self::Error> {
+        Ok(v.to_raw())
+    }
+}
+
+impl TryFromVal<Env, &Address> for RawVal {
+    type Error = ConversionError;
+
+    fn try_from_val(_env: &Env, v: &&Address) -> Result<Self, Self::Error> {
+        Ok(v.to_raw())
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFrom<&Address> for ScVal {
+    type Error = ConversionError;
+    fn try_from(v: &Address) -> Result<Self, Self::Error> {
+        ScVal::try_from_val(&v.env, &v.obj.to_raw())
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFrom<Address> for ScVal {
+    type Error = ConversionError;
+    fn try_from(v: Address) -> Result<Self, Self::Error> {
+        (&v).try_into()
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFromVal<Env, ScVal> for Address {
+    type Error = ConversionError;
+    fn try_from_val(env: &Env, val: &ScVal) -> Result<Self, Self::Error> {
+        use soroban_env_host::TryIntoVal;
+        <_ as TryFromVal<_, Object>>::try_from_val(
+            env,
+            &val.try_into_val(env).map_err(|_| ConversionError)?,
+        )
+    }
+}
+
+impl Address {
+    /// Ensures that this Address has authorized invocation of the current
+    /// contract with the provided arguments.
+    ///
+    /// ### Panics
+    ///
+    /// If the invocation is not authorized.
+    ///
+    /// During the on-chain execution the Soroban host will perform the needed
+    /// authentication (verify the signatures) and ensure the replay prevention.
+    /// The contracts don't need to perform this tasks.
+    ///
+    /// The arguments don't have to match the arguments of the contract
+    /// invocation. However, it's considered the best practice to have a
+    /// well-defined, deterministic and ledger-state-independent mapping between
+    /// the contract invocation arguments and `require_auth` arguments. This
+    /// will allow the contract callers to easily build the required signature
+    /// payloads and prevent potential authorization failures.
+    ///
+    /// When called in the tests, the `require_auth` calls are just recorded and
+    /// no signatures are required. In order to make sure that the contract
+    /// has indeed called `require_auth` for this Address with expected arguments
+    /// use `env.verify_top_authorization`.
```

Comment:
Usually the sections like Panic follow after the discussion paragraph.
```suggestion
    /// During the on-chain execution the Soroban host will perform the needed
    /// authentication (verify the signatures) and ensure the replay prevention.
    /// The contracts don't need to perform this tasks.
    ///
    /// The arguments don't have to match the arguments of the contract
    /// invocation. However, it's considered the best practice to have a
    /// well-defined, deterministic and ledger-state-independent mapping between
    /// the contract invocation arguments and `require_auth` arguments. This
    /// will allow the contract callers to easily build the required signature
    /// payloads and prevent potential authorization failures.
    ///
    /// When called in the tests, the `require_auth` calls are just recorded and
    /// no signatures are required. In order to make sure that the contract
    /// has indeed called `require_auth` for this Address with expected arguments
    /// use `env.verify_top_authorization`.
    ///
    /// ### Panics
    ///
    /// If the invocation is not authorized.
```

Reply:
Updated.